core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -67,9 +69,69 @@ public static int parseDate(String value)
         // - the required format is 'YYYY-MM-DD'
         // - all components should be unsigned numbers
         // https://github.com/trinodb/trino/issues/10677
+        OptionalInt days = tryParseDateYyyyMmDd(value);
+        if (days.isPresent()) {
+            return days.getAsInt();
+        }
         return toIntExact(TimeUnit.MILLISECONDS.toDays(DATE_FORMATTER.parseMillis(value)));
     }
 
+    /**
+     * Check if the date is in format yyyy-MM-dd, if yes, return the epoch days; otherwise
+     * return OptionalInt.empty(), indicating the parse failed, and should fall back to generic
+     * parsing. Mixing the format checking and parsing in one pass for better performance.
+     * @param value date string to be parsed
+     * @return the number of days since 1970-01-01
+     */
+    public static OptionalInt tryParseDateYyyyMmDd(String value)
+    {
+        if (value.length() != 10) {
+            return OptionalInt.empty();
```

Comment:
If you also validate that it is 4 digits, then hyphen, then two digits, then hyphen, then two digits, the following code will need to further checks, and be much simpler

Reply:
checking the format at the beginning was our initial implementation, but it turned out to be 25% slower, because that way we need to call String.charAt() more. We need to call it during checking and parsing. 
Though we can store the chars in an array when do format checking, but that also make codes a little longer, and GC pressure, since need to allocate an array each time.

below is the detailed benchmark result:
// mixing format check and parsing (current implementation)
Benchmark                      Mode  Cnt    Score    Error   Units
BenchmarkParseDate.parseDate  thrpt   10  742.748 ± 40.368  ops/ms

// check format at the begining
Benchmark                      Mode  Cnt    Score    Error   Units
BenchmarkParseDate.parseDate  thrpt   10  561.297 ± 29.083  ops/ms